# Software Coaching for Python
# Week 3: Pandas Dataframes - Homework

Instructor: Kang-Pyo Lee

In [ ]:
from google.colab import drive

drive.mount('/content/drive')
path = "/content/drive/Shared drives/PythonCoaching_2023Summer"
my_folder = "ljw33083418"     # *** REPLACE WITH YOUR FOLDER NAME ***
outcome_folder = f"{path}/{my_folder}/outcome"
classdata_folder = f"{path}/classdata"

Mounted at /content/drive


## Element Selection from a Dataframe

In [ ]:
import pandas as pd
pd.set_option('display.max_colwidth', 150)

df = pd.read_csv(f"{classdata_folder}/timeline_cnnbrk.csv", sep="\t")
df

status_id                      created_at  \
0     1312243714585235457  Sat Oct 03 04:10:56 +0000 2020   
1     1312217447752237056  Sat Oct 03 02:26:34 +0000 2020   
2     1312189828247691265  Sat Oct 03 00:36:49 +0000 2020   
3     1312188613015203842  Sat Oct 03 00:31:59 +0000 2020   
4     1312161028923244545  Fri Oct 02 22:42:22 +0000 2020   
...                   ...                             ...   
3195  1231426815996170243  Sun Feb 23 03:53:26 +0000 2020   
3196  1231402873617108992  Sun Feb 23 02:18:18 +0000 2020   
3197  1231385843723636739  Sun Feb 23 01:10:37 +0000 2020   
3198  1231338606863601668  Sat Feb 22 22:02:55 +0000 2020   
3199  1231307966487449600  Sat Feb 22 20:01:10 +0000 2020   

                                                                                                                                              text  \
0     President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA   
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
2     Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
4      President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for the next… https://t.co/Btb0FBkJLW   
...                                                                                                                                            ...   
3195                   Science Channel star Mad Mike Hughes died Saturday during an attempt to launch his homemade rocket… https://t.co/ezI1BUHstT   
3196  A person was killed in New Orleans when they were hit by a float during the Krewe of Endymion parade, the city says… https://t.co/wC9pSiSqTv   
3197  Sen. Bernie Sanders will win a decisive victory in Nevada's Democratic caucuses, CNN projects, solidifying his posi… https://t.co/Wk2nFjEiOw   
3198   Three people were killed and 18 injured after a bus rolled over on Interstate 15 near Fallbrook, California, north… https://t.co/UHqTVpV7Vm   
3199                                                     Nevada's Democratic caucuses are underway https://t.co/7O0yKSUzzI https://t.co/321iXf34lS   

      retweet_count  favorite_count  is_retweet retweet_created_at  
0              1050            3946           0                NaN  
1              2287            7232           0                NaN  
2               501            1559           0                NaN  
3              2831            5992           0                NaN  
4               249             841           0                NaN  
...             ...             ...         ...                ...  
3195            345             961           0                NaN  
3196            194             345           0                NaN  
3197            321            2058           0                NaN  
3198             97             305           0                NaN  
3199             29             196           0                NaN  

[3200 rows x 7 columns]

1\. Get the list of column labels of `df`.

In [ ]:
# Your answer here

# df의 행 label 가져오기
df.columns

Index(['status_id', 'created_at', 'text', 'retweet_count', 'favorite_count',
       'is_retweet', 'retweet_created_at'],
      dtype='object')

2\. Get the list of row index positions of `df`.

In [ ]:
# Your answer here

# df의 row index positions 가져오기
result = list(df.index) # df.index는 iterator (RangeIndex(start=0, stop=3200, step=1))
print(result)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221,

3\. Get the shape of `df`.

In [ ]:
# Your answer here
df.shape # (행 수, 열 수)

(3200, 7)

4\. Get the number of rows in `df`.

In [ ]:
# Your answer here
len(df) # dataframe의 길이 == 행 수

3200

5\. Select all values from the `created_at` column.

In [ ]:
# Your answer here
df["created_at"]

0       Sat Oct 03 04:10:56 +0000 2020
1       Sat Oct 03 02:26:34 +0000 2020
2       Sat Oct 03 00:36:49 +0000 2020
3       Sat Oct 03 00:31:59 +0000 2020
4       Fri Oct 02 22:42:22 +0000 2020
                     ...              
3195    Sun Feb 23 03:53:26 +0000 2020
3196    Sun Feb 23 02:18:18 +0000 2020
3197    Sun Feb 23 01:10:37 +0000 2020
3198    Sat Feb 22 22:02:55 +0000 2020
3199    Sat Feb 22 20:01:10 +0000 2020
Name: created_at, Length: 3200, dtype: object

6\. Select the first 3 rows from `df`.

In [ ]:
# Your answer here
df.head(3)

status_id                      created_at  \
0  1312243714585235457  Sat Oct 03 04:10:56 +0000 2020   
1  1312217447752237056  Sat Oct 03 02:26:34 +0000 2020   
2  1312189828247691265  Sat Oct 03 00:36:49 +0000 2020   

                                                                                                                                           text  \
0  President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA   
1  Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
2  Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp   

   retweet_count  favorite_count  is_retweet retweet_created_at  
0           1050            3946           0                NaN  
1           2287            7232           0                NaN  
2            501            1559           0                NaN

7\. Select the last 3 rows from `df`.

In [ ]:
# Your answer here
df.tail(3)

status_id                      created_at  \
3197  1231385843723636739  Sun Feb 23 01:10:37 +0000 2020   
3198  1231338606863601668  Sat Feb 22 22:02:55 +0000 2020   
3199  1231307966487449600  Sat Feb 22 20:01:10 +0000 2020   

                                                                                                                                              text  \
3197  Sen. Bernie Sanders will win a decisive victory in Nevada's Democratic caucuses, CNN projects, solidifying his posi… https://t.co/Wk2nFjEiOw   
3198   Three people were killed and 18 injured after a bus rolled over on Interstate 15 near Fallbrook, California, north… https://t.co/UHqTVpV7Vm   
3199                                                     Nevada's Democratic caucuses are underway https://t.co/7O0yKSUzzI https://t.co/321iXf34lS   

      retweet_count  favorite_count  is_retweet retweet_created_at  
3197            321            2058           0                NaN  
3198             97             305           0                NaN  
3199             29             196           0                NaN

8\. Select the element at row index position 999 and in the `text` column.

In [ ]:
# Your answer here
df["text"][999]

"US agents, local law enforcement and a locksmith's van entered the Chinese consulate in Houston days after it was o… https://t.co/25xHJptVgE"

9\. Select the element in the fifth row and the third column.

In [ ]:
# Your answer here
df.iloc[4][2]

'President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for the next… https://t.co/Btb0FBkJLW'

10\. Select all values from the last column, without specifying the column label.

In [ ]:
# Your answer here
df.iloc[:, -1] #Series

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
3195    NaN
3196    NaN
3197    NaN
3198    NaN
3199    NaN
Name: retweet_created_at, Length: 3200, dtype: object

11\. Select all rows in the last two columns, without specifying the column labels.

In [ ]:
# Your answer here
df.iloc[:, -2:] #Dataframe

is_retweet retweet_created_at
0              0                NaN
1              0                NaN
2              0                NaN
3              0                NaN
4              0                NaN
...          ...                ...
3195           0                NaN
3196           0                NaN
3197           0                NaN
3198           0                NaN
3199           0                NaN

[3200 rows x 2 columns]

12\. Select the first five rows in the last two columns, without specifying the column labels.

In [ ]:
# Your answer here
df.iloc[:5, -2:]

is_retweet retweet_created_at
0           0                NaN
1           0                NaN
2           0                NaN
3           0                NaN
4           0                NaN

13\. Select all rows from `df` with the `favorite_count` column value being strictly larger than 5000.

In [ ]:
# Your answer here
df[df.favorite_count > 5000]

status_id                      created_at  \
1     1312217447752237056  Sat Oct 03 02:26:34 +0000 2020   
3     1312188613015203842  Sat Oct 03 00:31:59 +0000 2020   
5     1312155447839780864  Fri Oct 02 22:20:12 +0000 2020   
17    1312067092980604928  Fri Oct 02 16:29:06 +0000 2020   
23    1311898474875703296  Fri Oct 02 05:19:04 +0000 2020   
...                   ...                             ...   
2933  1238646170307760128  Sat Mar 14 02:00:34 +0000 2020   
2951  1238286757264797697  Fri Mar 13 02:12:23 +0000 2020   
2953  1238259897743138816  Fri Mar 13 00:25:40 +0000 2020   
3009  1237395474371481600  Tue Mar 10 15:10:45 +0000 2020   
3036  1236432293906075651  Sat Mar 07 23:23:25 +0000 2020   

                                                                                                                                              text  \
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
5     The Biden campaign is pulling all negative ads after President Trump was diagnosed with Covid-19, a campaign aide s… https://t.co/iaxbszv9Zn   
17    Presidential nominee Joe Biden and his wife, Jill Biden, tested negative for coronavirus this morning, their doctor… https://t.co/ryLBhUByku   
23                          President Trump tweets that he and first lady Melania Trump have tested positive for Covid-19… https://t.co/didusJ091s   
...                                                                                                                                            ...   
2933  Google will not be publishing a national-scale website for coronavirus testing anytime soon, contrary to claims mad… https://t.co/xSz4KsBDXw   
2951  Canadian Prime Minister Justin Trudeau's wife Sophie has tested positive for coronavirus. The Prime Minister does n… https://t.co/lo6yZLeTsk   
2953  Disney is closing Walt Disney World, its flagship theme park resort in Florida, because of the coronavirus outbreak… https://t.co/2uWdlHQRaq   
3009                               Dick's Sporting Goods will stop selling guns at 440 more stores https://t.co/sYV3zPn7k3 https://t.co/OZpLeAAijM   
3036  A person who attended CPAC last week tested positive for coronavirus. Many officials, including Trump and Pence, we… https://t.co/31hH9vlDL7   

      retweet_count  favorite_count  is_retweet retweet_created_at  
1              2287            7232           0                NaN  
3              2831            5992           0                NaN  
5               948            8089           0                NaN  
17              894            7653           0                NaN  
23             2466            9103           0                NaN  
...             ...             ...         ...                ...  
2933           4348            9560           0                NaN  
2951           3413            6179           0                NaN  
2953           3172            5426           0                NaN  
3009            774            7854           0                NaN  
3036           4434           12796           0                NaN  

[195 rows x 7 columns]

14\. Select all rows from `df` with the `favorite_count` column value being between 5000 (inclusive) and 10000 (exclusive).

In [ ]:
# Your answer here
df[(df.favorite_count >= 5000) & (df.favorite_count < 10000)]

status_id                      created_at  \
1     1312217447752237056  Sat Oct 03 02:26:34 +0000 2020   
3     1312188613015203842  Sat Oct 03 00:31:59 +0000 2020   
5     1312155447839780864  Fri Oct 02 22:20:12 +0000 2020   
17    1312067092980604928  Fri Oct 02 16:29:06 +0000 2020   
23    1311898474875703296  Fri Oct 02 05:19:04 +0000 2020   
...                   ...                             ...   
2822  1240482077713334272  Thu Mar 19 03:35:49 +0000 2020   
2933  1238646170307760128  Sat Mar 14 02:00:34 +0000 2020   
2951  1238286757264797697  Fri Mar 13 02:12:23 +0000 2020   
2953  1238259897743138816  Fri Mar 13 00:25:40 +0000 2020   
3009  1237395474371481600  Tue Mar 10 15:10:45 +0000 2020   

                                                                                                                                              text  \
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
5     The Biden campaign is pulling all negative ads after President Trump was diagnosed with Covid-19, a campaign aide s… https://t.co/iaxbszv9Zn   
17    Presidential nominee Joe Biden and his wife, Jill Biden, tested negative for coronavirus this morning, their doctor… https://t.co/ryLBhUByku   
23                          President Trump tweets that he and first lady Melania Trump have tested positive for Covid-19… https://t.co/didusJ091s   
...                                                                                                                                            ...   
2822  For the first time since the novel coronavirus outbreak began late last year, China has reported no new local infec… https://t.co/KAHAmEbhi9   
2933  Google will not be publishing a national-scale website for coronavirus testing anytime soon, contrary to claims mad… https://t.co/xSz4KsBDXw   
2951  Canadian Prime Minister Justin Trudeau's wife Sophie has tested positive for coronavirus. The Prime Minister does n… https://t.co/lo6yZLeTsk   
2953  Disney is closing Walt Disney World, its flagship theme park resort in Florida, because of the coronavirus outbreak… https://t.co/2uWdlHQRaq   
3009                               Dick's Sporting Goods will stop selling guns at 440 more stores https://t.co/sYV3zPn7k3 https://t.co/OZpLeAAijM   

      retweet_count  favorite_count  is_retweet retweet_created_at  
1              2287            7232           0                NaN  
3              2831            5992           0                NaN  
5               948            8089           0                NaN  
17              894            7653           0                NaN  
23             2466            9103           0                NaN  
...             ...             ...         ...                ...  
2822           1438            7417           0                NaN  
2933           4348            9560           0                NaN  
2951           3413            6179           0                NaN  
2953           3172            5426           0                NaN  
3009            774            7854           0                NaN  

[141 rows x 7 columns]

15\. Create a random sample of 100 rows from `df` without duplicates.

In [ ]:
# Your answer here
df.sample(n=100, replace=False) #replace=False -> without duplicates

status_id                      created_at  \
1854  1267316369160364033  Mon Jun 01 04:45:42 +0000 2020   
3171  1232478690392465408  Wed Feb 26 01:33:12 +0000 2020   
2042  1263287031821676544  Thu May 21 01:54:33 +0000 2020   
1512  1274412301085478913  Sat Jun 20 18:42:24 +0000 2020   
1993  1264725107903168515  Mon May 25 01:08:57 +0000 2020   
...                   ...                             ...   
1095  1283746453882634242  Thu Jul 16 12:53:00 +0000 2020   
3074  1235045943524741120  Wed Mar 04 03:34:33 +0000 2020   
469   1299914048134361088  Sun Aug 30 03:37:14 +0000 2020   
753   1293289518024593410  Tue Aug 11 20:53:43 +0000 2020   
1023  1286070224819818498  Wed Jul 22 22:46:50 +0000 2020   

                                                                                                                                              text  \
1854             Minnesota officials believe there are white supremacists attending demonstrations https://t.co/e6EamKcLZt https://t.co/yOV0NdOZVZ   
3171    RT @CNNPolitics: Pete Buttigieg: There are seven white people on stage talking about racial justice #DemDebate https://t.co/DoHBlmq3q8 ht…   
2042  Peru has surpassed 100,000 confirmed coronavirus cases, the country's Ministry of Health says. Peru is second to Br… https://t.co/kIYCylOrvu   
1512                   Six campaign staffers working on President Trump's rally in Tulsa have tested positive for Covid-19 https://t.co/OfqzwcZRws   
1993                              Republican National Committee sues California to halt vote-by-mail for November election https://t.co/6wnew9dcKr   
...                                                                                                                                            ...   
1095            Another 1.3 million people filed first-time jobless claims last week, according to the Department of Labor https://t.co/SMKIcxD8an   
3074  RT @CNNPolitics: BREAKING: Former Vice President Joe Biden will win the Minnesota Democratic primary, CNN projects #CNNelection https://t.c…   
469   A suspect shot two police officers who were responding to a call in St. Louis, then barricaded himself in a home, a… https://t.co/KiSUZRiUsn   
753            The Pac-12 Conference CEO Group has unanimously voted to cancel the fall sports season, including football… https://t.co/E1Lyp27cNm   
1023       RT @CNNPolitics: JUST IN: House votes to remove Confederate statues from the US Capitol https://t.co/Dk5Ks7zSn5 https://t.co/7tr16pGlDv   

      retweet_count  favorite_count  is_retweet  \
1854           3071            9682           0   
3171            118               0           1   
2042            271             518           0   
1512           2072            6838           0   
1993            447            1494           0   
...             ...             ...         ...   
1095            890            1684           0   
3074            163               0           1   
469             264             672           0   
753             160             603           0   
1023            742               0           1   

                  retweet_created_at  
1854                             NaN  
3171  Wed Feb 26 01:31:41 +0000 2020  
2042                             NaN  
1512                             NaN  
1993                             NaN  
...                              ...  
1095                             NaN  
3074  Wed Mar 04 03:34:23 +0000 2020  
469                              NaN  
753                              NaN  
1023  Wed Jul 22 22:26:26 +0000 2020  

[100 rows x 7 columns]

## Dataframe Iteration & File Writing

Let's continue to use `df`.

1\. Select the first ten values from the `text` column of `df` and print the first 30 characters of each value concatenated by the string '...'. Rows in the printed output are separated by a new line. For example, the first three rows in the printed output look as follows.

President Donald Trump's campa...<br>
Former White House adviser Kel...<br>
Sen. Thom Tillis says he has t...<br>

In [ ]:
# Your answer here
for text in df[:10].text:
    print(text[:30] + "...")

President Donald Trump's campa...
Former White House adviser Kel...
Sen. Thom Tillis says he has t...
After Joe Biden's campaign sai...
President Trump has landed on ...
The Biden campaign is pulling ...
The White House says President...
President Trump is headed to W...
President Trump has had a feve...
Disgraced movie mogul Harvey W...


2\. Select the first ten rows from `df` and print the values of each row in the following format: 'created at: CREATED_AT, text: TEXT, retweet_count: RETWEET_COUNT', where CREATED_AT, TEXT, and RETWEET_COUNT are the values from the three columns `created_at`, `text`, and `retweet_count`, respectively. Rows are separated by two new lines. For example, the first three rows in the printed output look as follows.

created_at: Sat Oct 03 04:10:56 +0000 2020, text: President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA, retweet_count: 1050<br>
<br>
created_at: Sat Oct 03 02:26:34 +0000 2020, text: Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj, retweet_count: 2287<br>
<br>
created_at: Sat Oct 03 00:36:49 +0000 2020, text: Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp, retweet_count: 501<br>

In [ ]:
# Your answer here

# 10 row, 각각 위의 포멧에 맞춰서/ 각 row는 개행문자 두 개로 분리
for idx, row in df[:10].iterrows():
  print("created_at: " + str(row.created_at) + ", text: " + str(row.text) + "retweet_count: " + str(row.retweet_count), end="\n\n")

created_at: Sat Oct 03 04:10:56 +0000 2020, text: President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpAretweet_count: 1050

created_at: Sat Oct 03 02:26:34 +0000 2020, text: Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cjretweet_count: 2287

created_at: Sat Oct 03 00:36:49 +0000 2020, text: Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjpretweet_count: 501

created_at: Sat Oct 03 00:31:59 +0000 2020, text: After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzCretweet_count: 2831

created_at: Fri Oct 02 22:42:22 +0000 2020, text: President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for t

3\. Select all rows from `df` with the `is_retweet` column value being 0 and save the resulting dataframe in a CSV file named hw3_output.csv in the `outcome` folder such that the column separator is a tab and it contains the index of the dataframe.

In [ ]:
# Your answer here

# is_retweet 열의 값이 0인 모든 row를 hw3_output.csv로 저장/ separator는 tab, 각 row는 index를 포함해야함
df3 = df[df.is_retweet == 0]
df3.to_csv(f"{outcome_folder}/hw3_output.csv", sep="\t", index=True)

## Adding Columns to a Dataframe

Let's continue to use `df`.

1\. Add a new column `text_shortened` to `df` such that each value in the column is the first 30 characters of the `text` column value concatenated by the string '...'.  

In [ ]:
# Your answer here

# text_shortened 열 추가 (text열의 값 30글자 + '...'붙인거)
df["text_shortened"] = df.text.apply(lambda x: x[:30] + "...")

In [ ]:
# Check your answer here.
df[["text", "text_shortened"]]

text  \
0     President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA   
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
2     Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
4      President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for the next… https://t.co/Btb0FBkJLW   
...                                                                                                                                            ...   
3195                   Science Channel star Mad Mike Hughes died Saturday during an attempt to launch his homemade rocket… https://t.co/ezI1BUHstT   
3196  A person was killed in New Orleans when they were hit by a float during the Krewe of Endymion parade, the city says… https://t.co/wC9pSiSqTv   
3197  Sen. Bernie Sanders will win a decisive victory in Nevada's Democratic caucuses, CNN projects, solidifying his posi… https://t.co/Wk2nFjEiOw   
3198   Three people were killed and 18 injured after a bus rolled over on Interstate 15 near Fallbrook, California, north… https://t.co/UHqTVpV7Vm   
3199                                                     Nevada's Democratic caucuses are underway https://t.co/7O0yKSUzzI https://t.co/321iXf34lS   

                         text_shortened  
0     President Donald Trump's campa...  
1     Former White House adviser Kel...  
2     Sen. Thom Tillis says he has t...  
3     After Joe Biden's campaign sai...  
4     President Trump has landed on ...  
...                                 ...  
3195  Science Channel star Mad Mike ...  
3196  A person was killed in New Orl...  
3197  Sen. Bernie Sanders will win a...  
3198  Three people were killed and 1...  
3199  Nevada's Democratic caucuses a...  

[3200 rows x 2 columns]

2\. Add a new column `first_last_token` to `df` such that each column value is a tuple with two elements: the first element being the first token of the `text` column value and the second one being the last token (use the <b>split</b> method).

In [ ]:
# Your answer here

# first_last_token 열 추가 (text열의 값에서, (,) 형식으로)
def func(x):
  x = x.strip().split()
  return (x[0], x[-1])

df["first_last_token"] = df.text.apply(func)

In [ ]:
# Check your answer here.
df[["text", "first_last_token"]]

text  \
0     President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA   
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
2     Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
4      President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for the next… https://t.co/Btb0FBkJLW   
...                                                                                                                                            ...   
3195                   Science Channel star Mad Mike Hughes died Saturday during an attempt to launch his homemade rocket… https://t.co/ezI1BUHstT   
3196  A person was killed in New Orleans when they were hit by a float during the Krewe of Endymion parade, the city says… https://t.co/wC9pSiSqTv   
3197  Sen. Bernie Sanders will win a decisive victory in Nevada's Democratic caucuses, CNN projects, solidifying his posi… https://t.co/Wk2nFjEiOw   
3198   Three people were killed and 18 injured after a bus rolled over on Interstate 15 near Fallbrook, California, north… https://t.co/UHqTVpV7Vm   
3199                                                     Nevada's Democratic caucuses are underway https://t.co/7O0yKSUzzI https://t.co/321iXf34lS   

                          first_last_token  
0     (President, https://t.co/piwok6WXpA)  
1        (Former, https://t.co/DzHyaaO7Cj)  
2          (Sen., https://t.co/J3cnZUTEjp)  
3         (After, https://t.co/hArgCeKRzC)  
4     (President, https://t.co/Btb0FBkJLW)  
...                                    ...  
3195    (Science, https://t.co/ezI1BUHstT)  
3196          (A, https://t.co/wC9pSiSqTv)  
3197       (Sen., https://t.co/Wk2nFjEiOw)  
3198      (Three, https://t.co/UHqTVpV7Vm)  
3199   (Nevada's, https://t.co/321iXf34lS)  

[3200 rows x 2 columns]

3\. Add a new column `metrics` to `df` such that each column value is a tuple with three values from the three columns `retweet_count`, `favorite_count`, and `is_retweet`, respectively, in order.

In [ ]:
# Your answer here

# metrics 열 추가 (retweet_count, favorite_count, is_retweet) 형식으로
df["metrics"] = df.apply(lambda row: (row['retweet_count'], row['favorite_count'], row['is_retweet']), axis=1) #axis=1을 통해 행별로 적용

In [ ]:
# Check your answer here.
df[["retweet_count", "favorite_count", "is_retweet", "metrics"]]

retweet_count  favorite_count  is_retweet          metrics
0              1050            3946           0  (1050, 3946, 0)
1              2287            7232           0  (2287, 7232, 0)
2               501            1559           0   (501, 1559, 0)
3              2831            5992           0  (2831, 5992, 0)
4               249             841           0    (249, 841, 0)
...             ...             ...         ...              ...
3195            345             961           0    (345, 961, 0)
3196            194             345           0    (194, 345, 0)
3197            321            2058           0   (321, 2058, 0)
3198             97             305           0     (97, 305, 0)
3199             29             196           0     (29, 196, 0)

[3200 rows x 4 columns]

4\. Add a new column `num_mentions` to `df` such that each column value is the number of '@' symbols in the `text` column value.

In [ ]:
# Your answer here

# num_mentions 열 추가 (text열 값의 '@'의 개수)
df['num_mentions'] = df.text.apply(lambda x: x.count('@'))

In [ ]:
# Check your answer here.
df[["text", "num_mentions"]]

text  \
0     President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA   
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
2     Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
4      President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for the next… https://t.co/Btb0FBkJLW   
...                                                                                                                                            ...   
3195                   Science Channel star Mad Mike Hughes died Saturday during an attempt to launch his homemade rocket… https://t.co/ezI1BUHstT   
3196  A person was killed in New Orleans when they were hit by a float during the Krewe of Endymion parade, the city says… https://t.co/wC9pSiSqTv   
3197  Sen. Bernie Sanders will win a decisive victory in Nevada's Democratic caucuses, CNN projects, solidifying his posi… https://t.co/Wk2nFjEiOw   
3198   Three people were killed and 18 injured after a bus rolled over on Interstate 15 near Fallbrook, California, north… https://t.co/UHqTVpV7Vm   
3199                                                     Nevada's Democratic caucuses are underway https://t.co/7O0yKSUzzI https://t.co/321iXf34lS   

      num_mentions  
0                0  
1                0  
2                0  
3                0  
4                0  
...            ...  
3195             0  
3196             0  
3197             0  
3198             0  
3199             0  

[3200 rows x 2 columns]